src/containers/run.c

```diff
@@ -1026,6 +1008,39 @@ static inline int _avx2_run_container_cardinality(const run_container_t *run) {
     return sum;
 }
 
+ALLOW_UNALIGNED
+int _avx2_run_container_to_uint32_array(void *vout, const run_container_t *cont,
+                                        uint32_t base) {
+    int outpos = 0;
+    uint32_t *out = (uint32_t *)vout;
+
+    for (int i = 0; i < cont->n_runs; ++i) {
+        uint32_t run_start = base + cont->runs[i].value;
+        uint16_t le = cont->runs[i].length;
+        int j = 0;
+        __m256i run_start_v = _mm256_set1_epi32(run_start);
```

Comment:
What happens here if you instead of eagerly creating run_start_v, you only create it when the length is sufficiently large... say 8 or more.

This sort of tuning could make sure that your version if never much slower than our base version.

See what I mean?

Reply:
fixed

Reply:
This is not caused by creating run_start_v early. I suspect it's caused by the distribution of instructions. Such code does not appear to have the desired effect. Eventually adding LIKELY was the only thing that didn't cause performance degradation.


4.40/3.76/3.96
```
         if (le > 32) {
            __m256i run_start_v = _mm256_set1_epi32(run_start);
            // [8,8,8,8....]
            __m256i inc = _mm256_set1_epi32(8);
            // used for generate sequence:
            // [0, 1, 2, 3...], [8, 9, 10,...]
            __m256i delta = _mm256_setr_epi32(0, 1, 2, 3, 4, 5, 6, 7);
            for (j = 0; j + 8 <= le; j += 8) {
                __m256i val_v = _mm256_add_epi32(run_start_v, delta);
                _mm256_storeu_si256((__m256i *)(out + outpos), val_v);
                delta = _mm256_add_epi32(inc, delta);
                outpos += 8;
            }
        } else {
            for (; j <= le; ++j) {
                uint32_t val = run_start + j;
                memcpy(out + outpos, &val,
                       sizeof(uint32_t));  // should be compiled as a MOV on x64
                outpos++;
            }
        }
```

```

```